text/0000-constrained-types.md

```diff
@@ -0,0 +1,170 @@
+
+primitive Validated
+
+class val Error
+  let _errors: Array[String val] = _errors.create()
+
+  new create(e: (String val | None) = None) =>
+    match e
+    | let s: String val => _errors.push(s)
+    end
+
+  fun ref apply(e: String val) =>
+    _errors.push(e)
+
+  fun ref errors(): this->Array[String val] =>
+    _errors
+
+interface val Validator[T]
+  new val create()
+  fun apply(i: T): ValidationResult
+
+class val Valid[T: Any val, F: Validator[T]]
+  let _value: T val
+
+  new val _create(value: T val) =>
+    _value = value
+
+  fun val apply(): T val =>
+    _value
+
+primitive ValidConstructor[T: Any val, F: Validator[T] val]
+  fun apply(value: T): (Valid[T, F] | Error) =>
+    match F(value)
+    | Validated => Valid[T, F]._create(value)
+    | let e: Error => e
+    end
+```
+
+The library could be used thusly:
+
+```pony
+type LessThan10 is Valid[U64, LessThan10Validator]
+type MakeLessThan10 is ValidConstructor[U64, LessThan10Validator]
+
+primitive LessThan10Validator is Validator[U64]
+  fun apply(i: U64): ValidationResult =>
+    recover val
+      if i < 10 then
+        Validated
+      else
+        let s: String val = i.string() + " isn't less than 10"
+        Error(s)
+    end
+  end
+
+actor Main
+  new create(env: Env) =>
+    let prints = MakeLessThan10(U64(10))
+    match prints
+    | let p: LessThan10 => Foo(env.out, p).go()
+    | let e: Error =>
+      for s in e.errors().values() do
+        env.err.print(s)
+      end
+    end
+
+actor Foo
+  let out: OutStream
+  var left: U64
+
+  new create(out': OutStream, prints: LessThan10) =>
+    out = out'
+    left = prints()
+
+  be go() =>
+    if left > 0 then
+      out.print(left.string())
+      left = left - 1
+      go()
+    end
+```
+
+In our example usage code, we are creating a constrained type `LessThan10` that enforces that the value is between 0 and 9.
+
+Some key points from the design:
+
+## We can only validate immutable objects
+
+Validating a mutable item is pointless as it could change and go outside of our constraints after it has been validated. All validated items must be immutable.
+
+## `Error` is immutable
+
+We don't want to allow error mesages to be changed after validation is done. Because everything being validated is sendable, we can wrap an entire validator in a `recover` block and build up error messages on a `ref` Error before lifting to `val`.
+
+## Validators are not composable
+
+There's no safe way with the Pony type system that I can see to make a Validator composable. You can say for example that `SmallRange` is `GreaterThan5 & LessThan10` and then use a `SmallRange` where one a `LessThan10` is called for.
+
+# How We Teach This
+
+There's a few areas of teaching. One, the package documentation should have a couple of "here's how to use" examples with explanation as well as an explanation of why you would want to use the package instead of say checking a U64 repeatedly for being `< 10`.
```

Comment:
```suggestion
There's a few areas of teaching. One, the package documentation should have a couple of "here's how to use" examples with explanation as well as an explanation of why you would want to use the package instead of say checking a `U64` repeatedly for being `< 10`.
```